File: Remotion/Data/Linq.IntegrationTests.CSharp/LinqSamples101/GroupExistsInAnyAllContains.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Remotion.Data.Linq.IntegrationTests.Utilities;

namespace Remotion.Data.Linq.IntegrationTests.CSharp.LinqSamples101
{
  class GroupExistsInAnyAllContains:Executor
  {

    //This sample uses Any to return only Customers that have no Orders.")]
    public void LinqToSqlExists01 ()
    {
      var q =
          from c in db.Customers
          where !c.Orders.Any ()
          select c;

      serializer.Serialize (q);
    }

    //This sample uses Any to return only Categories that have at least one Discontinued product.")]
    public void LinqToSqlExists02 ()
    {
      var q =
          from c in db.Categories
          where c.Products.Any (p => p.Discontinued)
          select c;

      serializer.Serialize (q);
    }

    //This sample uses All to return Customers whom all of their orders have been shipped to their own city or whom have no orders.")]
    public void LinqToSqlExists03 ()
    {
      var q =
          from c in db.Customers
          where c.Orders.All (o => o.ShipCity == c.City)
          select c;

      serializer.Serialize (q);
    }

    //This sample uses Contain to find which Customer contains an order with OrderID 10248.")]
    public void LinqToSqlExists04 ()
    {
      var order = (from o in db.Orders
                   where o.OrderID == 10248
                   select o).First ();

      var q = db.Customers.Where (p => p.Orders.Contains (order)).ToList ();

      foreach (var cust in q)
      {
        foreach (var ord in cust.Orders)
        {
          serializer.Serialize (String.Format ("Customer {0} has OrderID {1}.", cust.CustomerID, ord.OrderID));
        }
      }
    }

    //This sample uses Contains to find customers whose city is Seattle, London, Paris or Vancouver.")]
    public void LinqToSqlExists05 ()
    {
      string[] cities = new string[] { "Seattle", "London", "Vancouver", "Paris" };
      var q = db.Customers.Where (p => cities.Contains (p.City)).ToList ();

      serializer.Serialize (q);
    }
  }
}

File: Remotion/Data/Linq.IntegrationTest.VisualBasic/LinqSamples101/GroupExistsInAnyAllContains.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.
Option Infer On
Option Strict On

Imports System.Collections.Generic
Imports System.Data
Imports System.Data.SqlClient
Imports System.IO
Imports System.Linq
Imports System.Linq.Expressions
Imports System.Reflection
Imports Remotion.Data.Linq.IntegrationTests.Utilities

Namespace LinqSamples101
  Public Class GroupExistsInAnyAllContains
    Inherits Executor
    '<Category("EXISTS/IN/ANY/ALL")> _
    ' <Title("Any - Simple")> _
    ' <Description("This sample uses the Any operator to return only Customers that have no Orders.")> _
    Public Sub LinqToSqlExists01()
      Dim custQuery = From cust In db.Customers _
                      Where Not cust.Orders.Any()

      serializer.Serialize(custQuery)
    End Sub

    '<Category("EXISTS/IN/ANY/ALL")> _
    '<Title("Any - Conditional")> _
    '<Description("This sample uses Any to return only Categories that have " & _
    '             "at least one Discontinued product.")> _
    Public Sub LinqToSqlExists02()
      Dim prodQuery = From cust In db.Categories _
                      Where (From prod In cust.Products Where prod.Discontinued).Any()

      serializer.Serialize(prodQuery)
    End Sub

    '<Category("EXISTS/IN/ANY/ALL")> _
    '<Title("All - Conditional")> _
    '<Description("This sample uses All to return Customers whom all of their orders " & _
    '             "have been shipped to their own city or whom have no orders.")> _
    Public Sub LinqToSqlExists03()
      Dim ordQuery = From cust In db.Customers _
                     Where cust.Orders.All(Function(ord) ord.ShipCity = cust.City)

      serializer.Serialize(ordQuery)
    End Sub

    '<Category("Exists/In/Any/All/Contains")> _
    '<Title("Contains - One Object")> _
    '<Description("This sample uses Contain to find which Customer contains an order with OrderID 10248.")> _
    Public Sub LinqToSqlExists04()

      Dim order = (From o In db.Orders _
                   Where o.OrderID = 10248).First()

      Dim q = db.Customers.Where(Function(p) p.Orders.Contains(order)).ToList()

      For Each cust In q
        For Each ord In cust.Orders

          serializer.Serialize(String.Format("Customer {0} has OrderID {1}.", _
                            cust.CustomerID, ord.OrderID))
        Next
      Next
    End Sub

    '<Category("Exists/In/Any/All/Contains")> _
    '<Title("Contains - Multiple values")> _
    '<Description("This sample uses Contains to find customers whose city is Seattle, London, Paris or Vancouver.")> _
    Public Sub LinqToSqlExists05()
      Dim cities = New String() {"Seattle", "London", "Vancouver", "Paris"}

      Dim q = db.Customers.Where(Function(p) cities.Contains(p.City)).ToList()

      serializer.Serialize(q)
    End Sub
  End Class
End Namespace
